File: Algorithms/Sorting_Algorithms/JS/MergeSort.js
function Merge(arr, left, middle, right){
    var left_size = middle + 1 - left;
    var right_size = right - left;

    L = new Array();
    R = new Array();

    for (let i = 0; i < left_size; i++)
        L[i] = arr[left + i];
    for (let j = 0; j < right_size; j++)
        R[i] = arr[middle + j];

    i = j = 0;
    k = left;

    while (i < left_size && j < right_size)
        arr[k++] = L[i] < R[j] ? L[i++] : R[j++];
    
    while (i < left_size)
        arr[k++] = L[i++];

    while (j < right_size)
        arr[k++] = R[j++];
}

function MergeSort(arr, left, righ){
    if (left < right){
        var middle = Math.round((righ - left) / 2);

        MergeSort(arr, left, righ);
        MergeSort(arr, left, righ);

        Merge(arr, left, middle, righ);
    }
}

function PrintArray(arr){
    sb = ""
    arr.forEach(element => {
        sb += ` ${element}`;
    });
     console.log(sb);
}

function main(){
    console.log("Merge Sort");
    array = [30, 40, 10, 60, 50, 80, 70, 20];
    console.log("Original vector: ");
    PrintArray(array);
    MergeSort(array);
    console.log("Sorted vector: ");
    PrintArray(array);
    console.log("finished...");
}

main();

File: Algorithms/Sorting_Algorithms/Java/MergeSort.java
public class MergeSort {

	private static void mergeSort(int[] array, int left, int right) {
		if (left < right) {
			
			int middle = (right + left ) / 2;
			
			mergeSort(array, left, middle);
			mergeSort(array, middle + 1, right);
			
			merge(array, left, middle, right);
		}
	}

	private static void merge(int[] array, int left, int middle, int right) {
		int lSize = middle - left + 1;
		int rSize = right - middle;
		
		int L[] = new int[lSize];
		int R[] = new int[rSize];
		
		for (int i = 0; i < lSize; i++)
			L[i] = array[left + i];
		for (int j = 0; j < rSize; j++)
			R[j] = array[middle + 1 + j];
		
		int	i = 0, j = 0;
		int k = left;
		
		while (i < lSize && j < rSize) 
			array[k++] = L[i] < R[j] ? L[i++] : R[j++];

		while(i < lSize)
			array[k++] = L[i++];
		
		while(j < rSize)
			array[k++] = R[j++];
	}
	
	public static void main(String [] args) {
		int array[] = {30, 10, 40, 20, 80, 60, 50, 70};
		int left, right;
		left = 0;
		right = array.length - 1;
		mergeSort(array, left, right);
		String sortedArray = java.util.Arrays.toString(array);
		System.out.println(sortedArray);
	}
}
